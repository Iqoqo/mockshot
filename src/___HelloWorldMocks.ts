export class HelloWorldMocks {
    static bar;

    static bar(mock: "no-valid-params" | "success"): any {
        switch (mock) {
            case "no-valid-params":
                return {
                  "error": "yes"
                }
            case "success":
                return {
                  "api": {
                    "data": "123"
                  }
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static shape;

    static shape(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "data": {
                    "_id": "1234"
                  },
                  "id": "1234sddd56",
                  "some": "value"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static foo;

    static foo(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "foo": "barr"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }
}
