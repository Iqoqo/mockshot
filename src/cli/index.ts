import { ApiGenerator } from "../generators/ApiGenerator";
import { GeneratorRunner } from "../GeneratorRunner";
import { SnapshotFetcher } from "./util";

export async function run(args: string[]) {
  const outputDir = args[0];

  const snapshots = SnapshotFetcher.getSnapshots();

  const generators = [new ApiGenerator()];

  const generatorRunner = new GeneratorRunner(outputDir, generators, snapshots);
  await generatorRunner.run();
}
