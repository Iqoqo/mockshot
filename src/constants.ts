export const MockshotTag = "[mockshot]";
export const Success = "success";
