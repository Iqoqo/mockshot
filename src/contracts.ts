export interface ISnapshot {
  key: string;
  packageName?: string;
  filePath: string;
  data: any;
}

export interface IApiSnapshot extends ISnapshot {
  data: IApiSnapData;
}

export interface IClassSnapshot extends ISnapshot {
  data: IClassSnapData;
}

export interface IApiSnapData extends IApiSnapDataBase {
  mockName: string;
}

export interface IClassSnapData {
  className: string;
  methodName: string;
  mockName: string;
  mock: any;
}

export interface IApiSnapDataBase {
  url: string;
  httpMethod: "post" | "get" | "put" | "delete" | "patch";
  mock: {
    statusCode: number;
    body: any;
    error?: string;
  };
}

export type MatcherReturn = {
  pass: boolean;
  message(): string | (() => string);
};
