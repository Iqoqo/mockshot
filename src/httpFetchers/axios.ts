class AxiosResponse extends BaseHttpFetcher {
  isValid(): boolean {
    return true;
  }
}
interface Axios {
  response: {
    config: {
      method: string;
      url: string;
    };
    status: number;
    data: any;
  };
}

function instanceOfA(fetchLibrary: any): fetchLibrary is Axios {
  return (
    "response" in object &&
    "response.config" in object &&
    response.config.method
  );
}
