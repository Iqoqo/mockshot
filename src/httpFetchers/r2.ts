interface R2 {
  response: {
    response: {
      opts: {
        method: string;
      };
    };
  };
}
