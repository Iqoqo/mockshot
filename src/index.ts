export * from "./matchers";
export * from "./generators";
export { run } from "./run";
