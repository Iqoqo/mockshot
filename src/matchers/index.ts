import stringify from 'json-stable-stringify';
import { toMatchApiMock } from "./ApiMockMatcher";
import { toMatchClassMock } from "./ClassMockMatcher";

declare global {
    namespace jest {
        // tslint:disable-next-line:interface-name
        interface Matchers<R> {
            toMatchClassMock(
                className: string,
                methodName: string,
                mockName: string,
                ignoredKeyPaths?: string[]
            ): R;
            toMatchApiMock(
                response: object,
                mockName: string,
            ): R;
        }
    }
}

expect.addSnapshotSerializer({
    test: val => val.mock,
    print: val => stringify(val, { space: '  ' })
});

const toMatchMock = toMatchClassMock

expect.extend({ toMatchMock, toMatchClassMock, toMatchApiMock });

export { toMatchMock, toMatchClassMock, toMatchApiMock };
