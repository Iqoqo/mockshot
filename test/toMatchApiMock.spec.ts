import "../src/matchers";
import axios from "axios";
import r2 from "r2";
const adapter = require("axios/lib/adapters/http");

const testUrl = "http://www.example.com";

describe("toMatchApiMock()", () => {
  it("Should work with axios module", async () => {
    const res = await axios.get(testUrl, { adapter });
    expect(res).toMatchApiMock("axios-success");
  });

  it("Should be rewritten (the test)", async () => {
    const res = await axios.get(testUrl, { adapter });
    expect(res).toMatchApiMock("axios-failure");
  });

  it("Should work with r2 module", async () => {
    const res = await r2.get(testUrl);
    await expect(res).toMatchApiMock("r2-success");
  });

  it("Should work with fetch module", async () => {
    const res = await fetch(testUrl);
    await expect(res).toMatchApiMock("fetch-success");
  });

  it("Should not work with unsupported response object (not of type axios, r2, fetch)", async () => {
    const res = {
      statttus: 200,
      my_url: testUrl
    };
    await expect(res).not.toMatchApiMock("failure-unsupported-response");
  });
});
