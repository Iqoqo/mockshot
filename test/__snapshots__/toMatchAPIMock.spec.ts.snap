// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchApiMock() Should fail: [mockshot] [[GET https://www.example.com failure]] 1`] = `
{
  "method": "GET",
  "mock": {
    "body": "<!doctype html>\\n<html>\\n<head>\\n    <title>Example Domain</title>\\n\\n    <meta charset=\\"utf-8\\" />\\n    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\" />\\n    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />\\n    <style type=\\"text/css\\">\\n    body {\\n        background-color: #f0f0f2;\\n        margin: 0;\\n        padding: 0;\\n        font-family: \\"Open Sans\\", \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;\\n        \\n    }\\n    div {\\n        width: 600px;\\n        margin: 5em auto;\\n        padding: 50px;\\n        background-color: #fff;\\n        border-radius: 1em;\\n    }\\n    a:link, a:visited {\\n        color: #38488f;\\n        text-decoration: none;\\n    }\\n    @media (max-width: 700px) {\\n        body {\\n            background-color: #fff;\\n        }\\n        div {\\n            width: auto;\\n            margin: 0 auto;\\n            border-radius: 0;\\n            padding: 1em;\\n        }\\n    }\\n    </style>    \\n</head>\\n\\n<body>\\n<div>\\n    <h1>Example Domain</h1>\\n    <p>This domain is established to be used for illustrative examples in documents. You may use this\\n    domain in examples without prior coordination or asking for permission.</p>\\n    <p><a href=\\"http://www.iana.org/domains/example\\">More information...</a></p>\\n</div>\\n</body>\\n</html>\\n",
    "status": 200
  },
  "mockName": "failure",
  "url": "https://www.example.com"
}
`;

exports[`toMatchApiMock() Should work: [mockshot] [[GET https://www.example.com success]] 1`] = `
{
  "method": "GET",
  "mock": {
    "body": "<!doctype html>\\n<html>\\n<head>\\n    <title>Example Domain</title>\\n\\n    <meta charset=\\"utf-8\\" />\\n    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\" />\\n    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />\\n    <style type=\\"text/css\\">\\n    body {\\n        background-color: #f0f0f2;\\n        margin: 0;\\n        padding: 0;\\n        font-family: \\"Open Sans\\", \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;\\n        \\n    }\\n    div {\\n        width: 600px;\\n        margin: 5em auto;\\n        padding: 50px;\\n        background-color: #fff;\\n        border-radius: 1em;\\n    }\\n    a:link, a:visited {\\n        color: #38488f;\\n        text-decoration: none;\\n    }\\n    @media (max-width: 700px) {\\n        body {\\n            background-color: #fff;\\n        }\\n        div {\\n            width: auto;\\n            margin: 0 auto;\\n            border-radius: 0;\\n            padding: 1em;\\n        }\\n    }\\n    </style>    \\n</head>\\n\\n<body>\\n<div>\\n    <h1>Example Domain</h1>\\n    <p>This domain is established to be used for illustrative examples in documents. You may use this\\n    domain in examples without prior coordination or asking for permission.</p>\\n    <p><a href=\\"http://www.iana.org/domains/example\\">More information...</a></p>\\n</div>\\n</body>\\n</html>\\n",
    "status": 200
  },
  "mockName": "success",
  "url": "https://www.example.com"
}
`;
